drivers: stepper: introduce rudimentary trapezoidal ramp control

jilaypandya opened the pull request:
# Trapezoidal Ramp Control for Stepper Motors:

The idea is to create a ramp lib that can be integrated in any driver, the step counting for the driver itself stays unaffected by integration a ramp lib

I have been using Shell to see how the ramping looks like.

Starting This PR to get some early feedback.

## Integration in driver
![Sequence Diagram](https://www.plantuml.com/plantuml/dpng/vLRRQXin47tVhnZwj0Rt1m8kX3GKGrERs2tjqw9enxQ8Lb8bhLt_VKVAthEEdL1AeG_6UvapcZbpPcODirvmOPZoyExQ62auyqAhTy0Sq7rYkWQVblC5h3mQWnPkjF9MIuas6hUtVSmzAmnSB-VHADpGTTzYBOhAHxnA4gKzWjxJ6HJ1j7eAiCAV9IgEa0ksjQm0jzC7eROWb4VBU8ZRmKPRA0YSFHo1gHnigK1Jl34jbo0z2lwJP3Phx1RrUqZlKTBL7k6EtXMe_0HciuJYWFSVEZbUrTZK4HrkaeJZzupYxl3GNWx1d4bUoc2U2uinyfUiTyWVGMo0IOiiFuQpAUy1ekBs2Ymwz1aHP39AtXDxMSuyIxbM6x6TGdp3TL5eDOLX-7M5gl-UOt8RmGGAfW_COGv46ZdEnHEzOx0sMQ78Z4amsWtj1obRzALLIOxF0akOz84BLycGYdc6Z-RyZEAsFejosJEPCivHLfL_YOBHsEfa6dTZmPBQajDbQWCBhAPlhajBS3SZaXRbI_ro1fFZwp5KyM3dmHaa3Jn7NLxEFZknrwyW5WfyRFok0LbGApOb23eiNTTjf_iIgRM1hzG04iZTbiHRwuxmZ9egfw4I3nmsN3muetJb0-EFwOJinau0EDspMrI1Kuo515DAXy_i6s8fjapXBvy5urQVFK7mqbeQHytRAImNg_bwllYItIq-tlPYhCa562rZ72TXJZxriRTCrVquZX2AspWCdzg_fl6K_qOc2WzGnzkEYtXg5plUNmHkUmjWJ3j3GYlrb2QFUSNI7eJVzG89i6g65DMC9pqNHcACQkWpwes6ZDWSQxEuABG9_qnEHZ2DYac0ljfUOHJqikW8AOu9saTXLtLsKR0GgjIbExkD29asgsGAUvIQ2t-yP2-zRM_SNN_FLklRvUhotL6QA3hUPbNp_p-iWRyovuBY37KPVTEKBihkildsRmwuRxUV5pVpzO_N3hWTVVV9mJBqkbu2TM-S01UC0g3VSCSTTjZT6sTLKKszIl4TJtzqbhquKNu3)

References:
https://www.boost.org/doc/libs/1_81_0/libs/safe_numerics/example/stepper-motor.pdf
Equation 13 in the above mentioned paper can also be seen in this app note.
https://ww1.microchip.com/downloads/en/Appnotes/doc8017.pdf (Thanks @andre-stefanov :) )

---
Known open points:

- [x] Integrate ramp control for `stepper_run` as well
- [x] use of sqrt in calculate_start_velocity, this function gets called once when the ramping starts from zero-velocity
         copied [isqrt](https://github.com/zephyrproject-rtos/zephyr/blob/0b90fd5adf1f01625412efadba4331b5041fb828/samples/kernel/metairq_dispatch/src/main.c#L76)
- [x] Implement unit tests
- [x] Introduce DT based selection of trapezoidal ramp, allowing inclusion of other ramps in future

---

Update: 23.04.2025
Thanks @andre-stefanov  for contributing unit-tests :)

jilaypandya committed:
drivers: stepper: introduce trapezoidal ramp control

Introduce rudimentary trapezoidal ramp control using following paper:
https://www.boost.org/doc/libs/1_81_0/libs/safe_numerics/example/stepper-motor.pdf

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

andre-stefanov commented on drivers/stepper/ramp/ramp_trapezoidal.c:
```diff
@@ -0,0 +1,191 @@
+		LOG_ERR("Error: Acceleration cannot be zero");
+		return -1;
+	}
+	/* Calculate the start velocity based on the acceleration */
+	/* Using the formula: t = sqrt(2 * d / a) */
+	/* where d = 1 step, a = acceleration */
+	uint64_t step_interval_in_ns = NSEC_PER_SEC / sqrt(acceleration * 2);
+	return step_interval_in_ns;
```
start velocity is currently assuming, that there is definitely an acceleration phase needed. But if the requested maximal velocity is lower than the velocity calculated by this formula, we should instantly switch into the `CONSTANT_SPEED` and use the requested velocity/interval and ignore the otherwise first acceleration interval.

this check should happen either in `calculate_start_velocity` or in `trapezoidal_reset_ramp_data`

jilaypandya replied:
i have introduced this check in `trapezoidal_reset_ramp_data`. Thanks :)

andre-stefanov commented on include/zephyr/drivers/stepper.h:
```diff
@@ -99,6 +99,18 @@ enum stepper_event {
+	/** Ramp acceleration in micro-steps per second squared */
+	uint32_t acceleration;
+	/** Ramp maximum velocity in micro-steps per second */
+	uint32_t max_velocity;
```
is `max_velocity` really needed? next step interval depends only on current interval and acceleration/deceleration values.

the actual limit will be given by the requested stepper interval

jilaypandya replied:
Trying to depict a trapezoidal ramp. I am also using max_velocity to check for valid_timing in move_by and move_to. But i'll give this a second thought for sure :).

jilaypandya replied:
Okay we can use that function, but we definitely need to have a better name for it though. set_microstep_interval atleast imo sounds like this interval will take place with immediate effect. Probably something like `set_target_step_interval` ?

jilaypandya commented on drivers/stepper/ramp/ramp.h:
```diff
@@ -0,0 +1,73 @@
+	/* Ramp state for deceleration */
+	DECELERATION,
+	/* Ramp state for forced deceleration */
+	FORCED_DECELERATION,
```
This name actually sounds much better, Thanks :)

jilaypandya commented on drivers/stepper/ramp/ramp.h:
```diff
@@ -0,0 +1,73 @@
+	/* Target position in µsteps */
+	uint32_t ramp_target_position;
+	/* Steps to be used while forced deceleration */
+	int32_t forced_deceleration_steps;
```
yes

jbehrensnx commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -109,6 +137,24 @@ static void update_coil_charge(const struct device *dev)
+		if (data->ramp_common.ramp_data.current_ramp_state ==
+		    STEPPER_RAMP_STATE_PRE_DECELERATION) {
+			if (data->step_count > 0) {
+				data->step_count++;
```
Why do you increment/decrement `data->step_count` here, this will cause incorrect positions in combination with the non ramp part of the function, which will still run.

jbehrensnx commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -268,6 +361,23 @@ static int gpio_stepper_set_microstep_interval(const struct device *dev,
+
+	K_SPINLOCK(&data->lock) {
+		data->ramp_common.ramp_profile.acceleration = ramp_profile->acceleration;
+		data->ramp_common.ramp_profile.max_velocity = ramp_profile->max_velocity;
```
You don't need `max_velocity`, you already have `set_microstep_interval`.

jilaypandya replied:
We need to find a consensus on how to rename that function, set_microstep_interval imo signifies more like a tick time, could be compared with for instance , can_set_bitrate.

dipakgmx replied:
I would say, once this PR is merged, the CONFIG_STEPPER_RAMP can be deleted and all motors should support ramping by default.

dipakgmx replied:
I would say, once this PR is merged, the CONFIG_STEPPER_RAMP can be deleted and all motors should support ramping by default.

jbehrensnx commented on drivers/stepper/ramp/Kconfig:
```diff
@@ -0,0 +1,18 @@
+
+if STEPPER_RAMP
+
+config STEPPER_RAMP_TRAPEZOIDAL
```
I would prefer a dt-binding instead to allow for a per stepper-controller decision on using ramps or not.

jbehrensnx commented on drivers/stepper/ramp/ramp_trapezoidal.c:
```diff
@@ -0,0 +1,187 @@
+	switch (ramp_data->current_ramp_state) {
+	case STEPPER_RAMP_STATE_ACCELERATION:
+		ramp_data->ramp_actual_position++;
+		new_step_interval_in_ns = current_step_interval_in_ns -
```
You cant always use the approximation. You will need to use the accurate algorithm using the sqrt for a number of steps after starting accelerating and a couple of steps before stopping when decelerating. This is because of the approximation error.

jbehrensnx commented:
My [PR 88342](https://github.com/zephyrproject-rtos/zephyr/pull/88342) already uses the same algorithm (through I forgot to mention the algorithm admittedly) for a step-dir implementation. Several decisions made during development for that are relevant for this PR as well, so it should be included in the discussion.

rruuaanng commented on dts/bindings/stepper/ramp-parameters.yaml:
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Collection of ramp controller bindings for stepper motors.
```
```suggestion
Collection of ramp controllers for stepper motors.
```

It's in the `bindings` directory.

jilaypandya replied:
done :)

dipakgmx commented:
This is not related to the changes made in this PR, but could the functions `move_to` & `move_by` call either of them, I mean, in `move_by`, call `move_to(current_pos+move_by_target)`? It seems to me that some bits of code are repeated in both the functions (or thats what I felt when reviewing)

jilaypandya commented:
> This is not related to the changes made in this PR, but could the functions `move_to` & `move_by` call either of them, I mean, in `move_by`, call `move_to(current_pos+move_by_target)`? It seems to me that some bits of code are repeated in both the functions (or thats what I felt when reviewing)

you are right, move_to should call move_by. When can i expect a patch from you :-)

jilaypandya commented on drivers/stepper/stepper_common.h:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @brief Stepper Motor run mode options
+ */
+enum stepper_run_mode {
```
So earlier it was in stepper.h however no api function actually requires this enum, it is used in drivers and now in ramp, since the run mode has to be passed to ramp as well.

jilaypandya commented on drivers/stepper/ramp/ramp.h:
```diff
@@ -0,0 +1,171 @@
+ */
+typedef void (*stepper_ramp_reset_ramp_runtime_data_t)(const struct stepper_ramp_config *config,
+						       struct stepper_ramp_common *ramp_common,
+						       const bool is_stepper_dir_changed,
```
yep that's correct :), Will do it in the next push.

jilaypandya commented on samples/drivers/stepper/generic/Kconfig:
```diff
@@ -15,6 +15,18 @@ config PING_PONG_N_REV
+	default 1000
+
+config RAMP_DECELERATION
+	int "Ramp deceleration micro-step/s²"
```
it actually shows up quite well in menuconfig :)

dipakgmx commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -98,6 +111,24 @@ static void update_coil_charge(const struct device *dev)
+			data->actual_position++;
+			return;
+		}
+		if (data->direction == STEPPER_DIRECTION_POSITIVE) {
```
Wouldn't an else be sufficient?

jilaypandya replied:
yep :-) I'll correct this in upcoming amend

andre-stefanov committed:
tests: added unit tests for stepper ramp api

testing following criteria:
- distance profile
- first acceleration interval
- next step intervals

Signed-off-by: Andre Stefanov <mail@andrestefanov.de>

jilaypandya commented on include/zephyr/drivers/stepper/stepper_ramp.h:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @brief Stepper ramp profile
+ */
+struct stepper_ramp_profile {
```
done moved to stepper.h

faxe1008 commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -129,6 +172,10 @@ static void update_direction_from_step_count(const struct device *dev)
 	} else {
 		LOG_ERR("Step count is zero");
 	}
+	if (data->direction != direction) {
```
Directly return this?

jilaypandya replied:
Done

faxe1008 commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -179,6 +240,23 @@ static void stepper_work_step_handler(struct k_work *work)
+	}
+	return false;
+#else
+	if (data->delay_in_ns > 0) {
```
Return this directly

jilaypandya replied:
Done

jilaypandya commented on drivers/stepper/ramp/ramp_trapezoidal.c:
```diff
@@ -0,0 +1,280 @@
+ *
+ * copied from samples/kernel/metairq_dispatch/src/main.c 0b90fd5
+ */
+static uint32_t isqrt(uint64_t n)
```
yep, I agree.

jbehrensnx commented:
As a general thing, in my experience, you need picosecond resolution for the ramp calculations, which are then converted to nanoseconds for the stepper controllers (but saving the picosecond values for the next step). Not sure if your 0.676 to counteract approximation inaccuracies reduces accuracy so much that this no longer matters.

Could could you maybe test if in an acceleration towards 50 or 200 steps/s in 1 sec actually takes 1 second? I remember during my development using this algorithm that I often got inaccuracies in the range of 20%, something that is not noticeable when using only the console without printing the times.

jilaypandya commented on tests/drivers/stepper/ramp/src/main.c:
```diff
@@ -0,0 +1,223 @@
+				     .distance_profile = DISTANCE_PROFILE(501, 0, 500),
+			     });
+
+	LOG_DBG("Test 200000 steps");
```
thanks. corrected

jilaypandya commented on tests/drivers/stepper/ramp/src/main.c:
```diff
@@ -0,0 +1,223 @@
+				     .distance_profile = DISTANCE_PROFILE(500, 0, 500),
+			     });
+
+	LOG_DBG("Test 100001 steps");
```
done

jilaypandya commented on tests/drivers/stepper/ramp/src/main.c:
```diff
@@ -0,0 +1,223 @@
+				     .distance_profile = DISTANCE_PROFILE(2, 0, 1),
+			     });
+
+	LOG_DBG("Test 100000 steps");
```
done

jilaypandya commented on drivers/stepper/ramp/ramp_trapezoidal.c:
```diff
@@ -0,0 +1,280 @@
+	case STEPPER_RAMP_STATE_PRE_DECELERATION:
+		if ((current_step_interval_in_ns >
+		     ramp_data->ramp_stop_step_interval_threshold_in_ns)) {
+			uint64_t start_velocity =
```
done

jilaypandya commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -98,6 +111,22 @@ static void update_coil_charge(const struct device *dev)
+#ifdef CONFIG_STEPPER_RAMP
+	if (data->ramp_common.ramp_runtime_data.current_ramp_state ==
+	    STEPPER_RAMP_STATE_PRE_DECELERATION) {
+		if (data->direction == STEPPER_DIRECTION_NEGATIVE) {
```
thanks. done
